cmd/k8srules/controller.go

```diff
@@ -0,0 +1,273 @@
+	kubepkg "github.com/amalgam8/amalgam8/pkg/kubernetes"
+)
+
+const (
+	// cacheResyncPeriod is the period in which we do a full resync of the rules cache.
+	cacheResyncPeriod = time.Duration(60) * time.Second
+)
+
+type controller struct {
+	// kubernetes REST client
+	client *rest.RESTClient
+
+	// rulesCache caches service rules resources from Kubernetes API
+	rulesCache      cache.Store
+	rulesController cache.ControllerInterface
+
+	// workqueue is used to queue and process events send from the cache controllers
+	workqueue *kubepkg.Workqueue
+
+	// validator validates rules against the rule schema
+	validator a8api.Validator
+
+	// namespace from which to sync endpoints/pods
+	namespace string
+
+	// stopChan for stop signals
+	stopChan chan struct{}
+}
+
+// New creates and starts the controller.
+func new(ctx context.Context, namespace string) (*controller, error) {
+	// If no namespace is specified, fallback to default namespace
+	if namespace == "" {
+		namespace = "default"
+	}
+
+	if err := initThirdPartyResource(); err != nil {
+		return nil, err
+	}
+
+	// Create the kubernetes client for the third party resource
+	client, err := createTPRClient()
+	if err != nil {
+		return nil, err
+	}
+
+	validator, err := a8api.NewValidator()
+	if err != nil {
+		return nil, err
+	}
+
+	workqueue := kubepkg.NewWorkqueue()
+	controller := &controller{
+		client:    client,
+		workqueue: workqueue,
+		validator: validator,
+		namespace: namespace,
+	}
+
+	controller.rulesCache, controller.rulesController = cache.NewInformer(
+		cache.NewListWatchFromClient(client, "rules", namespace, nil),
+		&kuberules.Rule{},
+		cacheResyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    workqueue.EnqueueingAddFunc(controller.addRule),
+			UpdateFunc: workqueue.EnqueueingUpdateFunc(controller.updateRule),
+		},
+	)
+
+	controller.workqueue.Start()
+	go controller.rulesController.Run(ctx.Done())
+
+	return controller, nil
+}
+
+// Stop synchronizing the rules.
+func (c *controller) Stop() error {
+	c.workqueue.Stop()
+
+	return nil
+}
+
+// initialize third party resource if it does not exist
+func initThirdPartyResource() error {
+	client, err := kubepkg.NewClient(kubepkg.Config{})
+	if err != nil {
+		return err
+	}
+
+	resName := fmt.Sprintf("%s.%s", kuberules.ResourceName, kuberules.ResourceGroupName)
+	_, err = client.Extensions().ThirdPartyResources().Get(resName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			tpr := &v1beta1.ThirdPartyResource{
+				ObjectMeta: v1.ObjectMeta{
+					Name: resName,
+				},
+				Versions: []v1beta1.APIVersion{
+					{Name: kuberules.ResourceVersion},
+				},
+				Description: kuberules.ResourceDescription,
+			}
+
+			result, err := client.Extensions().ThirdPartyResources().Create(tpr)
+			if err != nil {
+				return err
+			}
+			log.Printf("The ThridPartYresource %#v created\n", result)
+		} else {
+			return err
+		}
+	} else {
+		log.Printf("The ThirdPartyResource %s already exists\n", resName)
```

Comment:
Should we be using logrus instead of the standard logging library? Or alternately, maybe we should put `registry/utils/logging/logging.go` into `/pkg/logging/logging.go` and use that?